### PR TITLE
ASAN SSO fix

### PR DESCRIPTION
**Pull request by CaseyCarter**
Just testing right now, since I've never successfully run the ASAN tests locally.

### CODE REVIEW

**Review by miscco — APPROVED** [review]

**Comment by CaseyCarter** [comment]
Update: I still can't reproduce the bug.

Digging around, I've determined that the proposed fix shouldn't really fix anything: the copy (and copy-with-allocator) constructors that call `_Construct<_From_string>` initialize `_Mypair` in such a manner as to default-init `_String_val` which value-initializes `_Bx` calling the `_Bxty` default constructor which value-initializes `_Buf` resulting in all of the SSO elements being value-initialized. (Pause for breath.) Consequently, we shouldn't need to null-terminate _anything_ that constructs in SSO mode. My working theory is "compiler bug?" with the "?" since I haven't managed to repro this yet with 17.4 or 17.3.

**Comment by CaseyCarter** [comment]
> Update: I still can't reproduce the bug.

I have _finally_ figured out why this reproduces reliably with 17.3.0, and not at all with 17.4p1. 
```c++
#include <cassert>
#include <cstddef>
#include <cstring>
#include <string>
#include <new>

using namespace std;

int main() {
#ifdef _WIN64
    constexpr const char* text = "testtest";
    constexpr size_t n = 8;
#else
    constexpr const char* text = "test";
    constexpr size_t n = 4;
#endif

    string s0{text};
    assert(s0.c_str()[n] == '\0');

    alignas(string) unsigned char space[sizeof(string)];
    memset(space, 0xff, sizeof(space));

    string& s1 = *::new (&space) string{s0};
    assert(s1.c_str()[n] == '\0');

    s1.~string();
}
```
This line changed in between:

https://github.com/microsoft/STL/commit/1ea95df4d4c1bd85b6b973b134b8fd680caf9415#diff-47339b57101941628ad9f2c5369af8928f218a6716d304e14b7480dcc84392f6R2320

After the change, we value-init `_Bxty._Buf` in SSO mode so every element is a null terminator. Before, we value-init `_Bxty._Ptr` so the string is in long mode. Astute observers will realize that value-initializing the pointer member zeroes the first `sizeof(void*)` bytes of the `_Bxty`'s object representation, which is why the test string must be at least that many bytes long to trigger the bug.

TLDR: there is no bug to fix in 17.4.
